### ISSUE TITLE
Incorrect IDE0004 "Cast is redundant" when casting to a nullable reference type

### ISSUE THREAD

**Issue by ashmind** [dotnet/roslyn#42456]
**Version Used**: 3.3.1

**Steps to Reproduce**:

```csharp
var x = (string?)"a";
```

**Expected Behavior**:

IDE0004 is not shown for `(string?)` cast as it affects subsequent flow.
For example, in this case
```csharp
var x = (string?)"a";
x = null;
```
removing the cast will produce `CS8600` on the second line (which might even be an error if warn-as-error is enabled).

**Actual Behavior**:

IDE0004 "Cast is redundant" is shown for `(string?)` cast.

**Comment by sharwell** [comment]
Starting with 16.6 Preview 1, this is no longer a false positive.

/cc @jcouv
